### PR TITLE
Prevent close using dealer

### LINKED ISSUE #488: ObjectDisposedException occurs during Swarm<T>.SendMessageAsync()
```
System.ObjectDisposedException: Cannot access a disposed object.
Object name: 'System.Net.Sockets.Socket'.
  at System.Net.Sockets.Socket.ThrowIfDisposedAndClosed () [0x0001b] in <c9226b2f33b7421e85bb391f94075e49>:0 
  at System.Net.Sockets.Socket.Receive (System.Byte[] buffer, System.Int32 offset, System.Int32 size, System.Net.Sockets.SocketFlags socketFlags, System.Net.Sockets.SocketError& errorCode) [0x00000] in <c9226b2f33b7421e85bb391f94075e49>:0 
  at System.Net.Sockets.Socket.Receive (System.Byte[] buffer, System.Int32 offset, System.Int32 size, System.Net.Sockets.SocketFlags socketFlags) [0x00000] in <c9226b2f33b7421e85bb391f94075e49>:0 
  at System.Net.Sockets.Socket.Receive (System.Byte[] buffer) [0x00000] in <c9226b2f33b7421e85bb391f94075e49>:0 
  at NetMQ.Core.Utils.Signaler.Recv () [0x00000] in <21696b85a92a4a0eb8332ff57aebfd69>:0 
  at NetMQ.Core.Mailbox.TryRecv (System.Int32 timeout, NetMQ.Core.Command& command) [0x0001f] in <21696b85a92a4a0eb8332ff57aebfd69>:0 
  at NetMQ.Core.SocketBase.ProcessCommands (System.Int32 timeout, System.Boolean throttle) [0x00003] in <21696b85a92a4a0eb8332ff57aebfd69>:0 
  at NetMQ.Core.SocketBase.TrySend (NetMQ.Msg& msg, System.TimeSpan timeout, System.Boolean more) [0x00066] in <21696b85a92a4a0eb8332ff57aebfd69>:0 
  at NetMQ.NetMQSocket.TrySend (NetMQ.Msg& msg, System.TimeSpan timeout, System.Boolean more) [0x00000] in <21696b85a92a4a0eb8332ff57aebfd69>:0 
  at NetMQ.OutgoingSocketExtensions.TrySendFrame (NetMQ.IOutgoingSocket socket, System.TimeSpan timeout, System.Byte[] data, System.Int32 length, System.Boolean more) [0x00020] in <21696b85a92a4a0eb8332ff57aebfd69>:0 
  at NetMQ.OutgoingSocketExtensions.TrySendMultipartMessage (NetMQ.IOutgoingSocket socket, System.TimeSpan timeout, NetMQ.NetMQMessage message) [0x0005c] in <21696b85a92a4a0eb8332ff57aebfd69>:0 
  at Libplanet.Net.Swarm`1+<SendMessageAsync>d__82[T].MoveNext () [0x000d1] in <2695129d9ac147efb77d276ae19f66d4>:0 
```

/cc @limebell

### LINKED ISSUE #490: System.ArgumentOutOfRangeException: HashDigest<SHA256> must be 32 bytes, but 4 was given
```
9/3/2019 4:01:50 PM +09:00: A message has parsed: "Libplanet.Net.Messages.BlockHashes", from "0xC4c126e520c8298e04933A096ecAa28E177F3B2d.Unspecified/52.231.66.89:63187.1"
9/3/2019 4:01:50 PM +09:00: A raw message [frame count: 7] has received.
9/3/2019 4:01:50 PM +09:00: No blocks to require. Ignore BlockHashes.
9/3/2019 4:01:50 PM +09:00: An unexpected exception occured during ReceiveMessage(): System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation. ---> System.ArgumentOutOfRangeException: HashDigest<SHA256> must be 32 bytes, but 4 was given
Parameter name: hashDigest
  at Libplanet.HashDigest`1[T]..ctor (System.Byte[] hashDigest) [0x00053] in <2695129d9ac147efb77d276ae19f66d4>:0 
  at Libplanet.Net.NetMQFrameExtension.ConvertToHashDigest[T] (NetMQ.NetMQFrame frame) [0x00007] in <2695129d9ac147efb77d276ae19f66d4>:0 
  at Libplanet.Net.Messages.BlockHashes+<>c.<.ctor>b__1_0 (NetMQ.NetMQFrame f) [0x00000] in <2695129d9ac147efb77d276ae19f66d4>:0 
  at (wrapper delegate-invoke) System.Func`2[NetMQ.NetMQFrame,Libplanet.HashDigest`1[System.Security.Cryptography.SHA256]].invoke_TResult_T(NetMQ.NetMQFrame)
  at System.Linq.Enumerable+SelectListPartitionIterator`2[TSource,TResult].ToList () [0x0003c] in <3e4da02cf86b4fc686ed0ac61bffc210>:0 
  at System.Linq.Enumerable.ToList[TSource] (System.Collections.Generic.IEnumerable`1[T] source) [0x0001f] in <3e4da02cf86b4fc686ed0ac61bffc210>:0 
  at Libplanet.Net.Messages.BlockHashes..ctor (NetMQ.NetMQFrame[] frames) [0x00041] in <2695129d9ac147efb77d276ae19f66d4>:0 
  at (wrapper managed-to-native) System.Reflection.MonoCMethod.InternalInvoke(System.Reflection.MonoCMethod,object,object[],System.Exception&)
  at System.Reflection.MonoCMethod.InternalInvoke (System.Object obj, System.Object[] parameters) [0x00002] in <1f0c1ef1ad524c38bbc5536809c46b48>:0 
   --- End of inner exception stack trace ---
  at System.Reflection.MonoCMethod.InternalInvoke (System.Object obj, System.Object[] parameters) [0x00014] in <1f0c1ef1ad524c38bbc5536809c46b48>:0 
  at System.Reflection.MonoCMethod.DoInvoke (System.Object obj, System.Reflection.BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) [0x0007a] in <1f0c1ef1ad524c38bbc5536809c46b48>:0 
  at System.Reflection.MonoCMethod.Invoke (System.Reflection.BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) [0x00000] in <1f0c1ef1ad524c38bbc5536809c46b48>:0 
  at System.RuntimeType.CreateInstanceImpl (System.Reflection.BindingFlags bindingAttr, System.Reflection.Binder binder, System.Object[] args, System.Globalization.CultureInfo culture, System.Object[] activationAttributes, System.Threading.StackCrawlMark& stackMark) [0x00213] in <1f0c1ef1ad524c38bbc5536809c46b48>:0 
  at System.Activator.CreateInstance (System.Type type, System.Reflection.BindingFlags bindingAttr, System.Reflection.Binder binder, System.Object[] args, System.Globalization.CultureInfo culture, System.Object[] activationAttributes) [0x00095] in <1f0c1ef1ad524c38bbc5536809c46b48>:0 
  at System.Activator.CreateInstance (System.Type type, System.Object[] args) [0x00000] in <1f0c1ef1ad524c38bbc5536809c46b48>:0 
  at Libplanet.Net.Messages.Message.Parse (NetMQ.NetMQMessage raw, System.Boolean reply) [0x00161] in <2695129d9ac147efb77d276ae19f66d4>:0 
  at Libplanet.Net.Swarm`1[T].ReceiveMessage (System.Object sender, NetMQ.NetMQSocketEventArgs e) [0x00030] in <2695129d9ac147efb77d276ae19f66d4>:0 
System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation. ---> System.ArgumentOutOfRangeException: HashDigest<SHA256> must be 32 bytes, but 4 was given
Parameter name: hashDigest
  at Libplanet.HashDigest`1[T]..ctor (System.Byte[] hashDigest) [0x00053] in <2695129d9ac147efb77d276ae19f66d4>:0 
  at Libplanet.Net.NetMQFrameExtension.ConvertToHashDigest[T] (NetMQ.NetMQFrame frame) [0x00007] in <2695129d9ac147efb77d276ae19f66d4>:0 
  at Libplanet.Net.Messages.BlockHashes+<>c.<.ctor>b__1_0 (NetMQ.NetMQFrame f) [0x00000] in <2695129d9ac147efb77d276ae19f66d4>:0 
  at (wrapper delegate-invoke) System.Func`2[NetMQ.NetMQFrame,Libplanet.HashDigest`1[System.Security.Cryptography.SHA256]].invoke_TResult_T(NetMQ.NetMQFrame)
  at System.Linq.Enumerable+SelectListPartitionIterator`2[TSource,TResult].ToList () [0x0003c] in <3e4da02cf86b4fc686ed0ac61bffc210>:0 
  at System.Linq.Enumerable.ToList[TSource] (System.Collections.Generic.IEnumerable`1[T] source) [0x0001f] in <3e4da02cf86b4fc686ed0ac61bffc210>:0 
  at Libplanet.Net.Messages.BlockHashes..ctor (NetMQ.NetMQFrame[] frames) [0x00041] in <2695129d9ac147efb77d276ae19f66d4>:0 
  at (wrapper managed-to-native) System.Reflection.MonoCMethod.InternalInvoke(System.Reflection.MonoCMethod,object,object[],System.Exception&)
  at System.Reflection.MonoCMethod.InternalInvoke (System.Object obj, System.Object[] parameters) [0x00002] in <1f0c1ef1ad524c38bbc5536809c46b48>:0 
   --- End of inner exception stack trace ---
  at System.Reflection.MonoCMethod.InternalInvoke (System.Object obj, System.Object[] parameters) [0x00014] in <1f0c1ef1ad524c38bbc5536809c46b48>:0 
  at System.Reflection.MonoCMethod.DoInvoke (System.Object obj, System.Reflection.BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) [0x0007a] in <1f0c1ef1ad524c38bbc5536809c46b48>:0 
  at System.Reflection.MonoCMethod.Invoke (System.Reflection.BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) [0x00000] in <1f0c1ef1ad524c38bbc5536809c46b48>:0 
  at System.RuntimeType.CreateInstanceImpl (System.Reflection.BindingFlags bindingAttr, System.Reflection.Binder binder, System.Object[] args, System.Globalization.CultureInfo culture, System.Object[] activationAttributes, System.Threading.StackCrawlMark& stackMark) [0x00213] in <1f0c1ef1ad524c38bbc5536809c46b48>:0 
  at System.Activator.CreateInstance (System.Type type, System.Reflection.BindingFlags bindingAttr, System.Reflection.Binder binder, System.Object[] args, System.Globalization.CultureInfo culture, System.Object[] activationAttributes) [0x00095] in <1f0c1ef1ad524c38bbc5536809c46b48>:0 
  at System.Activator.CreateInstance (System.Type type, System.Object[] args) [0x00000] in <1f0c1ef1ad524c38bbc5536809c46b48>:0 
  at Libplanet.Net.Messages.Message.Parse (NetMQ.NetMQMessage raw, System.Boolean reply) [0x00161] in <2695129d9ac147efb77d276ae19f66d4>:0 
  at Libplanet.Net.Swarm`1[T].ReceiveMessage (System.Object sender, NetMQ.NetMQSocketEventArgs e) [0x00030] in <2695129d9ac147efb77d276ae19f66d4>:0 
9/3/2019 4:01:50 PM +09:00: A raw message [frame count: 5] has received.
9/3/2019 4:01:50 PM +09:00: An unexpected exception occured during ReceiveMessage(): System.OverflowException: Arithmetic operation resulted in an overflow.
  at Libplanet.Net.Messages.Message.Parse (NetMQ.NetMQMessage raw, System.Boolean reply) [0x00029] in <2695129d9ac147efb77d276ae19f66d4>:0 
  at Libplanet.Net.Swarm`1[T].ReceiveMessage (System.Object sender, NetMQ.NetMQSocketEventArgs e) [0x00030] in <2695129d9ac147efb77d276ae19f66d4>:0 
System.OverflowException: Arithmetic operation resulted in an overflow.
  at Libplanet.Net.Messages.Message.Parse (NetMQ.NetMQMessage raw, System.Boolean reply) [0x00029] in <2695129d9ac147efb77d276ae19f66d4>:0 
  at Libplanet.Net.Swarm`1[T].ReceiveMessage (System.Object sender, NetMQ.NetMQSocketEventArgs e) [0x00030] in <2695129d9ac147efb77d276ae19f66d4>:0 
9/3/2019 4:01:51 PM +09:00: A raw message [frame count: 10] has received.
9/3/2019 4:01:51 PM +09:00: A message has parsed: "Libplanet.Net.Messages.TxIds", from "0xC4c126e520c8298e04933A096ecAa28E177F3B2d.Unspecified/52.231.66.89:63187.1"
9/3/2019 4:01:51 PM +09:00: Trying to fetch txs...
```

### LINKED ISSUE #491: System.OverflowException: Arithmetic operation resulted in an overflow.
```
9/3/2019 4:01:50 PM +09:00: A raw message [frame count: 5] has received.
9/3/2019 4:01:50 PM +09:00: An unexpected exception occured during ReceiveMessage(): System.OverflowException: Arithmetic operation resulted in an overflow.
  at Libplanet.Net.Messages.Message.Parse (NetMQ.NetMQMessage raw, System.Boolean reply) [0x00029] in <2695129d9ac147efb77d276ae19f66d4>:0 
  at Libplanet.Net.Swarm`1[T].ReceiveMessage (System.Object sender, NetMQ.NetMQSocketEventArgs e) [0x00030] in <2695129d9ac147efb77d276ae19f66d4>:0 
System.OverflowException: Arithmetic operation resulted in an overflow.
  at Libplanet.Net.Messages.Message.Parse (NetMQ.NetMQMessage raw, System.Boolean reply) [0x00029] in <2695129d9ac147efb77d276ae19f66d4>:0 
  at Libplanet.Net.Swarm`1[T].ReceiveMessage (System.Object sender, NetMQ.NetMQSocketEventArgs e) [0x00030] in <2695129d9ac147efb77d276ae19f66d4>:0 
9/3/2019 4:01:51 PM +09:00: A raw message [frame count: 10] has received.
```

### LINKED ISSUE #492: Libplanet.Net.InvalidMessageException: The message signature is invalid
```
9/3/2019 4:44:08 PM +09:00: A raw message [frame count: 6] has received.
9/3/2019 4:44:08 PM +09:00: Could not parse NetMQMessage properly; ignore: Libplanet.Net.InvalidMessageException: The message signature is invalid
  at Libplanet.Net.Messages.Message.Parse (NetMQ.NetMQMessage raw, System.Boolean reply) [0x001b0] in <2695129d9ac147efb77d276ae19f66d4>:0 
  at Libplanet.Net.Swarm`1[T].ReceiveMessage (System.Object sender, NetMQ.NetMQSocketEventArgs e) [0x00030] in <2695129d9ac147efb77d276ae19f66d4>:0 
Libplanet.Net.InvalidMessageException: The message signature is invalid
  at Libplanet.Net.Messages.Message.Parse (NetMQ.NetMQMessage raw, System.Boolean reply) [0x001b0] in <2695129d9ac147efb77d276ae19f66d4>:0 
  at Libplanet.Net.Swarm`1[T].ReceiveMessage (System.Object sender, NetMQ.NetMQSocketEventArgs e) [0x00030] in <2695129d9ac147efb77d276ae19f66d4>:0 
9/3/2019 4:44:08 PM +09:00: A raw message [frame count: 5] has received.
```

### PR DESCRIPTION
**Pull request by limebell**
This resolves #488, resolves #490, resolves #491, resolves #492 (maybe?)

### CODE REVIEW

File: Libplanet/Net/Swarm.cs [hunk 1]
```diff
@@ -823,22 +836,26 @@ internal async Task SendMessageAsync(BoundPeer peer, Message message)
                 _logger.Error(e, "Different version received.");
                 throw;
             }
-            catch (TimeoutException e)
+            catch (TimeoutException)
             {
-                _logger.Error(e, "Timeout occurred during SendMessageWithReplyAsync().");
+#pragma warning disable MEN002
```
Comment (dahlia) [inline Libplanet/Net/Swarm.cs]:
```suggestion
```

File: Libplanet/Net/Swarm.cs [hunk 2]
```diff
@@ -823,22 +836,26 @@ internal async Task SendMessageAsync(BoundPeer peer, Message message)
...
+                _logger.Debug(
+                    $"Timeout occurred during {nameof(SendMessageWithReplyAsync)}() after {timeout}.");
+#pragma warning restore MEN002
```
Comment (dahlia) [inline Libplanet/Net/Swarm.cs]:
```suggestion
```

File: Libplanet/Net/Swarm.cs [hunk 3]
```diff
@@ -823,22 +836,26 @@ internal async Task SendMessageAsync(BoundPeer peer, Message message)
                 _logger.Error(e, "Different version received.");
                 throw;
             }
-            catch (TimeoutException e)
+            catch (TimeoutException)
             {
-                _logger.Error(e, "Timeout occurred during SendMessageWithReplyAsync().");
+#pragma warning disable MEN002
+                _logger.Debug(
+                    $"Timeout occurred during {nameof(SendMessageWithReplyAsync)}() after {timeout}.");
```
Comment (dahlia) [inline Libplanet/Net/Swarm.cs]:
```suggestion
                    $"Timeout occurred during {nameof(SendMessageWithReplyAsync)}() after {{0}}.",
                    timeout);
```

File: Libplanet/Net/Swarm.cs [hunk 4]
```diff
@@ -823,22 +836,26 @@ internal async Task SendMessageAsync(BoundPeer peer, Message message)
...
             }
             catch (TaskCanceledException)
             {
-                _logger.Debug("Task canceled during SendMessageWithReplyAsync().");
+                _logger.Debug($"Task canceled during {nameof(SendMessageWithReplyAsync)}().");
                 throw;
             }
             catch (Exception e)
             {
+#pragma warning disable MEN002
```
Comment (dahlia) [inline Libplanet/Net/Swarm.cs]:
```suggestion
```

File: Libplanet/Net/Swarm.cs [hunk 5]
```diff
@@ -823,22 +836,26 @@ internal async Task SendMessageAsync(BoundPeer peer, Message message)
...
                 _logger.Error(
                     e,
-                    "An unexpected exception occurred during SendMessageWithReplyAsync(). {0}",
-                    e);
+                    $"An unexpected exception occurred during {nameof(SendMessageWithReplyAsync)}(). {e}");
+#pragma warning restore MEN002
```
Comment (dahlia) [inline Libplanet/Net/Swarm.cs]:
```suggestion
```

File: Libplanet/Net/Swarm.cs [hunk 6]
```diff
@@ -823,22 +836,26 @@ internal async Task SendMessageAsync(BoundPeer peer, Message message)
...
                 throw;
             }
             catch (Exception e)
             {
+#pragma warning disable MEN002
                 _logger.Error(
                     e,
-                    "An unexpected exception occurred during SendMessageWithReplyAsync(). {0}",
-                    e);
+                    $"An unexpected exception occurred during {nameof(SendMessageWithReplyAsync)}(). {e}");
```
Comment (dahlia) [inline Libplanet/Net/Swarm.cs]:
```suggestion
                    "An unexpected exception occurred during " +
                    $"{nameof(SendMessageWithReplyAsync)}(): {{0}}",
                    e);
```

File: Libplanet/Net/Swarm.cs [hunk 7]
```diff
@@ -1440,10 +1454,10 @@ private void BroadcastTxIds(IEnumerable<TxId> txIds)
                     _logger.Debug("Append complete.");
                 }
             }
-            catch (TimeoutException e)
+            catch (TimeoutException)
             {
                 // As we have more chances, ignore this.
-                _logger.Debug(e, $"Timeout occurred during {nameof(ProcessBlockHashes)}(): {e}");
+                _logger.Debug($"Timeout occurred during {nameof(ProcessBlockHashes)}()");
```
Comment (dahlia) [inline Libplanet/Net/Swarm.cs]:
```suggestion
                _logger.Debug($"Timeout occurred during {nameof(ProcessBlockHashes)}().");
```

File: Libplanet/Net/Swarm.cs [hunk 8]
```diff
@@ -1607,6 +1621,12 @@ CancellationToken cancellationToken
                     _logger.Debug("Swapping complete");
                 }
 
+                var blockHashes =
+                    blocks.Aggregate(string.Empty, (current, block) =>
+                        current + $"[{block.Hash.ToString()}]");
+                _logger.Debug($"Re-broadcast {nameof(BlockHashes)} with {blocks.Count} blocks " +
```
Comment (dahlia) [inline Libplanet/Net/Swarm.cs]:
```suggestion
                _logger.Debug($"Re-broadcast {nameof(BlockHashes)} with {{0}} blocks " +
```

File: Libplanet/Net/Swarm.cs [hunk 9]
```diff
@@ -1607,6 +1621,12 @@ CancellationToken cancellationToken
...
+                              $"which are {blockHashes}.");
```
Comment (dahlia) [inline Libplanet/Net/Swarm.cs]:
```suggestion
                              "which are {{1}}.",
                              blocks.Count,
                              blockHashes);
```

**Comment by moreal** [comment]
It seems to resolve #393, too.

**Comment by dahlia** [comment]
@limebell This pull request needs to be rebased on the current master.

**Pull request merged**